pkg/sql/alter_database.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
+
+import (
+	"context"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+)
+
+type alterDatabaseOwnerNode struct {
+	n    *tree.AlterDatabaseOwner
+	desc *sqlbase.MutableDatabaseDescriptor
+}
+
+// AlterDatabaseOwner transforms a tree.AlterDatabaseOwner into a plan node.
+func (p *planner) AlterDatabaseOwner(
+	ctx context.Context, n *tree.AlterDatabaseOwner,
+) (planNode, error) {
+	immutableDbDesc, err := p.ResolveUncachedDatabaseByName(ctx, n.Name.String(), true /*required*/)
+	if err != nil {
+		return nil, err
+	}
+
+	dbDesc := sqlbase.NewMutableExistingDatabaseDescriptor(*immutableDbDesc.DatabaseDesc())
+	privs := dbDesc.GetPrivileges()
+
+	// Make sure the newOwner exists.
+	roleExists, err := p.RoleExists(ctx, n.Owner)
```

Comment:
This code is duplicated -- we should be able to extract this logic into a function that operates on `catalog.MutableDescriptor` right?

Reply:
Can you point this to where this code is duplicated.

Also I'm unsure about the second part, why does it have to operate on `catalog.MutableDescriptor`

Reply:
Im thinking that some portions are duplicated with `p.alterTableOwner`. I'm `MutableDescriptor` because we are able to change the owners of schemas, databases, tables and types right? Those are all under the `MutableDescriptor` interface.

Reply:
Oh do you mean the checks? I'm not convinced it's actually worth deduplicating here, even though some checks are similar, there are small checks that differ inbetween for example alter table has to check for create on schema whereas alter db doesn't. 

They are short enough that I'm not sure it's worth abstracting.

Reply:
Here's the code that I think is duplicated: 
```
	// Make sure the newOwner exists.
	roleExists, err := p.RoleExists(ctx, newOwner)
	if err != nil {
		return false, err
	}
	if !roleExists {
		return false, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", newOwner)
	}

	// Make sure the user has ownership on the table
	// and not just create privilege.
	if !n.hasOwnership {
		return false, pgerror.Newf(pgcode.InsufficientPrivilege,
			"must be owner of table %s", tree.Name(desc.GetName()))
	}

	// Requirements from PG:
	// To alter the owner, you must also be a direct or indirect member of the
	// new owning role, and that role must have CREATE privilege on the
	// table's schema.
	memberOf, err := p.MemberOfWithAdminOption(ctx, privs.Owner)
	if err != nil {
		return false, err
	}
	if _, ok := memberOf[newOwner]; !ok {
		return false, pgerror.Newf(
			pgcode.InsufficientPrivilege, "must be member of role %q", newOwner)
	}

	// If the owner we want to set to is the current owner, do a no-op.
	if newOwner == privs.Owner {
		return false, nil
	}

	privs.SetOwner(newOwner)
```